Preflights report

jgruica opened the pull request.

salahalsaleh commented on kotsadm/pkg/handlers/deploy.go:
```diff
@@ -37,8 +41,21 @@ type UpdateUndeployResultRequest struct {
 	IsError bool   `json:"isError"`
 }
 
+type DeployAppVersionResponse struct {
```
```suggestion
type DeployAppVersionRequest struct {
```

salahalsaleh commented on kotsadm/pkg/handlers/deploy.go:
```diff
@@ -231,3 +306,43 @@ func (h *Handler) UpdateUndeployResult(w http.ResponseWriter, r *http.Request) {
+		return errors.Wrap(err, "failed to read")
+	}
+
+	fmt.Printf("%s\n", b)
```
clean up

salahalsaleh commented on kotsadm/pkg/handlers/deploy.go:
```diff
@@ -231,3 +306,43 @@ func (h *Handler) UpdateUndeployResult(w http.ResponseWriter, r *http.Request) {
+func updatePreflightsReportToReplicatedApp(license *kotsv1beta1.License, appSlug string, clusterID string, sequence int, skipPreflights bool, installStatus string) error {
+	urlValues := url.Values{}
+
+	sequenceToStr := strconv.Itoa(sequence)
```
```suggestion
	sequenceStr := fmt.Sprintf("%d", sequence)
```

salahalsaleh commented on kotsadm/pkg/handlers/deploy.go:
```diff
@@ -55,17 +69,29 @@ func (h *Handler) DeployAppVersion(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
-		err = errors.Wrap(err, "failed to list downstreams for app")
-		logger.Error(err)
-		w.WriteHeader(http.StatusInternalServerError)
```
why remove these? we should write this header

salahalsaleh commented on kotsadm/pkg/handlers/deploy.go:
```diff
@@ -55,17 +69,29 @@ func (h *Handler) DeployAppVersion(w http.ResponseWriter, r *http.Request) {
 	} else if len(downstreams) == 0 {
-		err = errors.New("no downstreams for app")
-		logger.Error(err)
-		w.WriteHeader(http.StatusInternalServerError)
```
same here

salahalsaleh commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,43 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+	foundApp, err := store.GetStore().GetAppFromSlug(appSlug)
+	if err != nil {
+		logger.Error(err)
+		w.WriteHeader(500)
```
```suggestion
		w.WriteHeader(http.StatusInternalServerError)
```

salahalsaleh commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,43 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+	license, err := store.GetStore().GetLatestLicenseForApp(foundApp.ID)
+	if err != nil {
+		logger.Error(err)
+		w.WriteHeader(500)
```
```suggestion
		w.WriteHeader(http.StatusInternalServerError)
```

salahalsaleh commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,43 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	JSON(w, 200, struct{}{})
```
```suggestion
	JSON(w, http.StatusOK, struct{}{})
```

salahalsaleh commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,43 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	clusterID := downstreams[0].ClusterID
```
we should discuss this on standup, because i think we're trying to move away from using this id

salahalsaleh commented on kotsadm/pkg/reporting/reporting.go:
```diff
@@ -128,3 +134,83 @@ func getK8sVersion() (string, error) {
+	return nil
+}
+
+func PreflightInfoThreadSend(appID string, appSlug string, sequence int, isSkipPreflights bool, isFailedPreflight bool) error {
```
this function never returns (it will keep running forever in some cases). i suggest checking the application deployment status in the go routine (thread), and only report when the application deployment process is done and has a status like "failed, "deployed", etc...)

salahalsaleh commented on kotsadm/pkg/online/online.go:
```diff
@@ -229,6 +229,17 @@ func CreateAppFromOnline(pendingApp *types.PendingApp, upstreamURI string, isAut
 			return nil, errors.Wrap(err, "failed to start preflights")
 		}
+		isUpdate := false
+		if err := reporting.SendPreflightInfo(pendingApp.ID, int(newSequence), skipPreflights, isUpdate); err != nil {
```
this seems to be duplicated by the one on line 239, since the function is called regardless if preflights are skipped or not

divolgin commented on kotsadm/pkg/handlers/deploy.go:
```diff
@@ -66,6 +80,22 @@ func (h *Handler) DeployAppVersion(w http.ResponseWriter, r *http.Request) {
+	if request.IsSkipPreflights {
+		isUpdate := false
+		if err := reporting.SendPreflightInfo(a.ID, int(sequence), request.IsSkipPreflights, isUpdate); err != nil {
+			logger.Error(errors.Wrap(err, "failed to start preflight thread"))
```
log.Debug error and continue with deployment

divolgin commented on kotsadm/pkg/handlers/deploy.go:
```diff
@@ -66,6 +80,22 @@ func (h *Handler) DeployAppVersion(w http.ResponseWriter, r *http.Request) {
+	if request.ContinueWithFailedPreflights && !request.IsSkipPreflights {
+		isUpdate := true
+		if err := reporting.SendPreflightInfo(a.ID, int(sequence), request.IsSkipPreflights, isUpdate); err != nil {
+			logger.Error(errors.Wrap(err, "failed to start preflight thread"))
```
log.Debug error and continue with deployment

divolgin commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,37 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(204)
 }
+
+func (h *Handler) SkipPreflights(w http.ResponseWriter, r *http.Request) {
```
Function is named `SkipPreflights` but I don't see it skipping preflights.  It only sends a report.

divolgin commented on kotsadm/web/src/components/PreflightResultPage.jsx:
```diff
@@ -247,14 +251,38 @@ class PreflightResultPage extends Component {
+      method: "POST",
+    })
+      .then((res) => {
+        if (res.status === 200) {
```
`if` and `else` are identical here.  so is `catch`.
`sendSkipPreflightsData` doesn't need to wait for this call to complete before navigating to the app page.

divolgin commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,37 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+	foundApp, err := store.GetStore().GetAppFromSlug(appSlug)
+	if err != nil {
+		logger.Debugf("failed to get app from slug", err)
+		return
```
none of these write any status to response before returning

divolgin commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,37 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+		logger.Debugf("failed to get downstreams for app", err)
+		return
+	} else if len(downstreams) == 0 {
+		logger.Debugf("no downstreams for app", err)
```
err is nil here

divolgin commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,37 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+
+	downstreams, err := store.GetStore().ListDownstreamsForApp(foundApp.ID)
+	if err != nil {
+		logger.Debugf("failed to get downstreams for app", err)
```
missing %v for err

divolgin commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,37 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+
+	foundApp, err := store.GetStore().GetAppFromSlug(appSlug)
+	if err != nil {
+		logger.Debugf("failed to get app from slug", err)
```
missing %v for err

divolgin commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,37 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+
+	license, err := store.GetStore().GetLatestLicenseForApp(foundApp.ID)
+	if err != nil {
+		logger.Debugf("failed to get latest license for app", err)
```
missing %v for err

divolgin commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,37 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+	clusterID := downstreams[0].ClusterID
+
+	if err := reporting.SendPreflightsReportToReplicatedApp(license, foundApp.ID, clusterID, 0, true, ""); err != nil {
+		logger.Debugf("failed to send preflights data to replicated app", err)
```
missing %v for err

divolgin commented on kotsadm/pkg/handlers/update.go:
```diff
@@ -58,6 +59,14 @@ func (h *Handler) AppUpdateCheck(w http.ResponseWriter, r *http.Request) {
+		err = reporting.SendPreflightInfo(foundApp.ID, int(foundApp.CurrentSequence), skipPreflights, isUpdate)
+		if err != nil {
+			logger.Debugf("failed to update preflights reports", err)
+			return
```
don't return here

divolgin commented on kotsadm/pkg/online/online.go:
```diff
@@ -230,5 +231,11 @@ func CreateAppFromOnline(pendingApp *types.PendingApp, upstreamURI string, isAut
 
+	isUpdate := false
+	if err := reporting.SendPreflightInfo(pendingApp.ID, int(newSequence), skipPreflights, isUpdate); err != nil {
+		logger.Error(errors.Wrap(err, "failed to start preflight thread"))
```
log debug and don't return here

divolgin commented on kotsadm/pkg/reporting/reporting.go:
```diff
@@ -1 +1,97 @@
+		return errors.Errorf("Unexpected status code %d", resp.StatusCode)
+	}
+
+	_, err = ioutil.ReadAll(resp.Body)
```
no need to read body here if it's not used.

divolgin commented on kotsadm/pkg/reporting/reporting.go:
```diff
@@ -1 +1,97 @@
+func SendPreflightInfo(appID string, sequence int, isSkipPreflights bool, isUpdate bool) error {
+	license, err := store.GetStore().GetLatestLicenseForApp(appID)
+	if err != nil {
+		logger.Error(errors.Wrap(err, "failed to find license for app"))
```
return error, don't log

divolgin commented on kotsadm/pkg/reporting/reporting.go:
```diff
@@ -1 +1,97 @@
+
+	downstreams, err := store.GetStore().ListDownstreamsForApp(appID)
+	if err != nil {
+		logger.Error(errors.Wrap(err, "failed to list downstreams for app"))
```
return error, don't log

divolgin commented on kotsadm/pkg/reporting/reporting.go:
```diff
@@ -1 +1,97 @@
+		logger.Error(errors.Wrap(err, "failed to list downstreams for app"))
+		return nil
+	} else if len(downstreams) == 0 {
+		logger.Error(errors.Wrap(err, "no downstreams for app"))
```
err is nil here, so Wrap won't work.  Also return error, don't log

divolgin commented on kotsadm/pkg/reporting/reporting.go:
```diff
@@ -1 +1,97 @@
+			}
+			if currentVersion.Status != "" && currentVersion.Status != "deploying" {
+				if err := SendPreflightsReportToReplicatedApp(license, appID, clusterID, sequence, isSkipPreflights, currentVersion.Status); err != nil {
+					logger.Error(errors.Wrap(err, "failed to send preflights data to replicated app"))
```
log debug

divolgin commented on kotsadm/pkg/reporting/reporting.go:
```diff
@@ -1 +1,97 @@
+		}()
+	} else {
+		if err := SendPreflightsReportToReplicatedApp(license, appID, clusterID, sequence, isSkipPreflights, ""); err != nil {
+			logger.Error(errors.Wrap(err, "failed to send preflights data to replicated app"))
```
return error, don't log

divolgin commented on kotsadm/pkg/handlers/deploy.go:
```diff
@@ -66,6 +80,22 @@ func (h *Handler) DeployAppVersion(w http.ResponseWriter, r *http.Request) {
 
+	if request.IsSkipPreflights {
+		isUpdate := false
+		if err := reporting.SendPreflightInfo(a.ID, int(sequence), request.IsSkipPreflights, isUpdate); err != nil {
```
call this as a go routine so it doesn't block deployment

divolgin commented on kotsadm/pkg/handlers/deploy.go:
```diff
@@ -66,6 +80,22 @@ func (h *Handler) DeployAppVersion(w http.ResponseWriter, r *http.Request) {
+
+	if request.ContinueWithFailedPreflights && !request.IsSkipPreflights {
+		isUpdate := true
+		if err := reporting.SendPreflightInfo(a.ID, int(sequence), request.IsSkipPreflights, isUpdate); err != nil {
```
call this as a go routine so it doesn't block deployment

divolgin commented on kotsadm/pkg/handlers/preflight.go:
```diff
@@ -279,3 +280,37 @@ func (h *Handler) PostPreflightStatus(w http.ResponseWriter, r *http.Request) {
+
+	clusterID := downstreams[0].ClusterID
+
+	if err := reporting.SendPreflightsReportToReplicatedApp(license, foundApp.ID, clusterID, 0, true, ""); err != nil {
```
call this as a go routine so it doesn't block deployment

divolgin commented on kotsadm/pkg/handlers/update.go:
```diff
@@ -58,6 +59,14 @@ func (h *Handler) AppUpdateCheck(w http.ResponseWriter, r *http.Request) {
 
+		// preflights reporting
+		isUpdate := true
+		err = reporting.SendPreflightInfo(foundApp.ID, int(foundApp.CurrentSequence), skipPreflights, isUpdate)
```
call this as a go routine so it doesn't block update check

divolgin commented on kotsadm/pkg/online/online.go:
```diff
@@ -230,5 +231,11 @@ func CreateAppFromOnline(pendingApp *types.PendingApp, upstreamURI string, isAut
 	}
 
+	isUpdate := false
+	if err := reporting.SendPreflightInfo(pendingApp.ID, int(newSequence), skipPreflights, isUpdate); err != nil {
```
call this as a go routine so it doesn't block installation

divolgin reviewed (changes requested):
Comments, but also no reporting functions should be called in airgapped installation.  At least the ones in the deploy function break that rule.

jgruica commented:
> Comments, but also no reporting functions should be called in airgapped installation. At least the ones in the deploy function break that rule.

@divolgin addressed comments in latest commit
